OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs

```diff
@@ -0,0 +1,90 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.TS.Traits
+{
+	[Desc("Play an animation when a unit exits or blocks the exit after production finished.")]
+	class WithProductionDoorOverlayInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>, Requires<BuildingInfo>
+	{
+		public readonly string Sequence = "idle-door";
+		public readonly string BuildSequence = "build-door";
+
+		public object Create(ActorInitializer init) { return new WithProductionDoorOverlay(init, this); }
+
+		public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer init, RenderSpritesInfo rs, string image, int facings, PaletteReference p)
+		{
+			var anim = new Animation(init.World, image, () => 0);
+			anim.PlayRepeating(Sequence);
+
+			var bi = init.Actor.Traits.Get<BuildingInfo>();
+			var offset = FootprintUtils.CenterOffset(init.World, bi).Y + 512; // Additional 512 units move from center -> top of cell
+			yield return new SpriteActorPreview(anim, WVec.Zero, offset, p, rs.Scale);
+		}
+	}
+
+	class WithProductionDoorOverlay : INotifyBuildComplete, ITick, INotifyProduction, INotifySold, INotifyDamageStateChanged
+	{
+		readonly WithProductionDoorOverlayInfo info;
+		readonly RenderSprites renderSprites;
+		readonly Animation door;
+
+		bool isOpen;
+		CPos openExit;
+		bool buildComplete;
+
+		public WithProductionDoorOverlay(ActorInitializer init, WithProductionDoorOverlayInfo info)
+		{
+			this.info = info;
+			renderSprites = init.Self.Trait<RenderSprites>();
+			door = new Animation(init.World, renderSprites.GetImage(init.Self));
+			var buildingInfo = init.Self.Info.Traits.Get<BuildingInfo>();
+
+			var offset = FootprintUtils.CenterOffset(init.World, buildingInfo).Y + 512;
+			renderSprites.Add("door_overlay_{0}".F(info.Sequence), new AnimationWithOffset(door, null, () => !buildComplete, offset));
+		}
+
+		public void BuildingComplete(Actor self)
+		{
+			door.Play(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.Sequence));
+			buildComplete = true;
+		}
+
+		public void Tick(Actor self)
+		{
+			if (isOpen && !self.World.ActorMap.GetUnitsAt(openExit).Any(a => a != self))
+			{
+				isOpen = false;
```

Comment:
That won't work. It will then only close when a new actor is produced which is glitchy in a different kind of way and I can't get that fixed so I will keep the legacy behavior for now.